### ISSUE TITLE
Clean up Penumbra-defined ABCI Events to use Protobufs

### ISSUE THREAD

**Issue by hdevalence** [penumbra-zone/penumbra#3460]
**Is your feature request related to a problem? Please describe.**

Currently, Penumbra's events are defined in an ad-hoc way: each component has an `event.rs` file with Rust functions returning ABCI events. These are constructed on-demand and emitted with the state's `.record()` method, then never used again.

There is no schema, no format definition, and all of the fields are encoded in different ways: sometimes hex, sometimes bech32, sometimes strings, etc. This makes it practically very difficult for any external tooling to consume, and the only reason we haven't noticed is that we haven't yet made a serious attempt at trying.

The SDK had a similar problem, and belatedly migrated to "typed events" defined using Protobufs. We should adopt a similar system, and migrate all of the Penumbra-defined events to be Protobuf-defined types, before external parties start trying to build on top of them

**Describe the solution you'd like**

- [ ] Define a system where events are defined by Protobuf messages. 

  The way the SDK does this can be [seen here](https://github.com/cosmos/cosmos-sdk/blob/main/types/events.go#L89-L156).  For our purposes, it doesn't seem necessary to implement a full domain-type/proto-type split like other proto messages. We don't have any Rust code that consumes the events, so we don't need to parse them, only generate them -- but we should ensure that tooling in _other_ languages can easily parse the event back to a Proto-generated type.  

  One possibility would be to define a trait in the proto crate, something like
  ```rust
  pub trait ProtoEvent: Message + Name {
      fn into_event(&self) -> abci::Event {
          // Use the `Name` trait to extract the proto type name
          // Serialize to JSON and munge strings as in Go code
      }
      fn from_event(event: &abci::Event) -> Result<Self> {
          // Use the `Name` trait to check the event type
          // Assemble k/v pairs into JSON, deserialize to proto
      }
  }
  ```
  
  Then, change the `StateWriteProto` trait in the `penumbra-proto` crate to add a method that does proto-event recording:
  ```rust
  pub trait StateWriteProto: StateWrite + Send + Sync {
      // ... (existing methods) ...
      
      /// Records a protobuf message as a typed ABCI event.
      fn record_proto<E: Message + Name>(&mut self, proto_event: E) {
          self.record(proto_event.into_event());
      }
  }
  ```
  
  Then we can still have the `event::foo()` event creation methods that exist now, but rather than constructing an ABCI event, they can return the proto message, which can be fed into `state.record_proto(...)`. This also means we have an easy way to search the codebase for uses of typed events vs. other events.

- [ ] #3504

  We should then replace all existing events defined in the Penumbra application with Protobuf-based events. These events should be defined in the Proto module corresponding to the component that emits the event. Currently, there are only 12 events currently defined (VS Code search `fn` in files named `event.rs`), so this shouldn't be a huge lift.
  
  Since any Protobuf message could in principle be encoded as an event this way, we should adopt a naming convention to distinguish messages defining events from messages defining other data. One possibility would be to name all of the event messages `Event*`. Using a prefix rather than a suffix means that all of the events will be grouped together when messages are sorted alphabetically.

**Additional context**

This work should be done before building tooling that consumes any of the existing events (in particular, before building Penumbra-specific parts of a block explorer), so we won't have to rewrite that tooling as part of making the change.

**Comment by ejmg** [comment]
The issue hasn't been updated yet but I'm going to work on this issue with the goal of making next week's test-net release.

**Comment by hdevalence** [comment]
I moved porting the existing event.rs files into its own issue, #3504, so it can be non-blocking on the testnet release. (It's slightly cleaner if it happened before, but we shouldn't block on it).
